core/src/KokkosExp_InterOp.hpp

```diff
@@ -0,0 +1,172 @@
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_CORE_EXP_INTEROP_HPP
+#define KOKKOS_CORE_EXP_INTEROP_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+#include <Kokkos_Layout.hpp>
+#include <Kokkos_MemoryTraits.hpp>
+#include <Kokkos_View.hpp>
+#include <Kokkos_DynRankView.hpp>
+#include <impl/Kokkos_Utilities.hpp>
+#include <type_traits>
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+template <typename... T>
+using type_list = Kokkos::Impl::type_list<T...>;
+
+// ------------------------------------------------------------------ //
+//  concat_type_list combines types in multiple type_lists
+
+// default definition
+template <typename... T>
+struct concat_type_list {
+  using type = type_list<T...>;
+};
+
+// append to type_list
+template <typename... T, typename... Tail>
+struct concat_type_list<type_list<T...>, Tail...>
+    : concat_type_list<T..., Tail...> {};
+
+// combine consecutive type_lists
+template <typename... T, typename... U, typename... Tail>
+struct concat_type_list<type_list<T...>, type_list<U...>, Tail...>
+    : concat_type_list<type_list<T..., U...>, Tail...> {};
```

Comment:
> This is conflating types and type_list.

Well, not really. It was more of a design decision to linearly expand the template instantiation instead of using recursive template instantiation like:

```cpp
    using type = concat_type_list_t<TL0, concat_type_list_t<TLs...>>;
```

I may be wrong but I _think_ my method will compile faster. We can update it to @nliber's recommendation but I'd be interested in y'all's thoughts w.r.t. that.

Reply:
Yeah, the provided definition for `concat_type_list` didn't work for various reasons. I'm going with this definition, which doesn't support passing non-type-lists but produces half as many `concat_type_list` instantiations:

```cpp
// forward declaration
template <typename... T>
struct concat_type_list;

// alias
template <typename... T>
using concat_type_list_t = typename concat_type_list<T...>::type;

// final instantiation
template <typename... T>
struct concat_type_list<type_list<T...>> {
  using type = type_list<T...>;
};

// combine consecutive type_lists
template <typename... T, typename... U, typename... Tail>
struct concat_type_list<type_list<T...>, type_list<U...>, Tail...>
    : concat_type_list<type_list<T..., U...>, Tail...> {};
```

Note, using `template <typename...> class TL, typename... Ts` + specializing on `TL<Ts...>`, e.g.:

```cpp
template <template <typename...> class TL, typename... Ts, typename... Us> 
struct concat_type_list<TL<Ts...>, TL<Us...>> {
    using type = TL<Ts..., Us...>;
};
```

causes many problems by treating `Kokkos::MemoryTraits<...>`, `Kokkos::Device<ExecT, MemT>`, etc. as "type-lists".

Reply:
>  I would have expected that the template arguments are required to be `type_list`

New definition requires this